### PR TITLE
Refactor NPO to use primitives with tfp.distribution

### PR DESCRIPTION
**Pull request by ahtsan**
Refactor NPO to use primitives with tfp.distribution.

### CODE REVIEW

**Comment by ahtsan** [comment]
Benchmark results will be uploaded once it's done.

File: src/garage/tf/algos/batch_polopt.py [hunk 1]
```diff
@@ -241,13 +241,8 @@ def process_samples(self, itr, paths):
 
         lengths = np.asarray([v.sum() for v in valids])
 
-        ent = np.sum(self.policy.distribution.entropy(agent_infos) *
-                     valids) / np.sum(valids)
-
         self.episode_reward_mean.extend(undiscounted_returns)
 
-        tabular.record('Entropy', ent)
```
Comment (ryanjulian) [inline src/garage/tf/algos/batch_polopt.py:249]: why?

File: src/garage/tf/algos/npo2.py [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+
+                # Maximize E[surrogate objective] by minimizing
+                # -E_t[surrogate objective]
+                if self.policy.recurrent:
+                    # filter valid obj
+                    obj = tf.boolean_mask(obj, i.valid_var)
+                loss = -tf.reduce_mean(obj)
+
+            # Diagnostic functions
+            self._f_policy_kl = compile_function(
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:367]: can you use tf.Session.make_callable instead?

File: src/garage/tf/algos/__init__.py [hunk 3]
```diff
@@ -5,14 +5,17 @@
 from garage.tf.algos.dqn import DQN
 from garage.tf.algos.erwr import ERWR
 from garage.tf.algos.npo import NPO
+from garage.tf.algos.npo2 import NPO2
 from garage.tf.algos.ppo import PPO
+from garage.tf.algos.ppo2 import PPO2
```
Comment (ryanjulian) [inline src/garage/tf/algos/__init__.py:10]: i would prefer to just delete the old one, as long as tests still pass and benchmarks look good

File: src/garage/tf/algos/npo2.py [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+                obs_var = tf.compat.v1.placeholder(
+                    tf.float32,
+                    shape=[None, None, observation_space.flat_dim],
+                    name='obs')
+            else:
+                obs_var = observation_space.to_tf_placeholder(name='obs',
+                                                              batch_dims=2)
+            action_var = action_space.to_tf_placeholder(name='action',
+                                                        batch_dims=2)
+            reward_var = new_tensor(name='reward', ndim=2, dtype=tf.float32)
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:223]: can you switch these out for tf.compat.v1.placeholder?

File: src/garage/tf/algos/__init__.py (hunk 3 above)
Comment (ahtsan) [reply src/garage/tf/algos/__init__.py:10]: Sure

File: src/garage/tf/algos/npo2.py (hunk 4 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:223]: Sure

File: src/garage/tf/algos/npo2.py (hunk 2 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:367]: Sure

File: src/garage/tf/algos/npo2.py [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+              use_neg_logli_entropy=False
+
+    """
+
+    def __init__(self,
+                 env_spec,
+                 policy,
+                 baseline,
+                 scope=None,
+                 max_path_length=500,
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:86]: this is a dangerous default -- 100 is more typical

File: src/garage/tf/algos/npo2.py [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+                                          stop_entropy_gradient,
+                                          use_neg_logli_entropy,
+                                          policy_ent_coeff)
+
+        if pg_loss not in ['vanilla', 'surrogate', 'surrogate_clip']:
+            raise ValueError('Invalid pg_loss')
+
+        with self._name_scope:
+            self._optimizer = optimizer(**optimizer_args)
+            self._lr_clip_range = float(lr_clip_range)
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:126]: i don't think the following 3 need to be in a name scope...

File: src/garage/tf/algos/npo2.py [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+                for k, shape in self.policy.state_info_specs
+            }
+            policy_state_info_vars_list = [
+                policy_state_info_vars[k] for k in self.policy.state_info_keys
+            ]
+
+        if self.policy.recurrent:
+            _obs_var = obs_var
+            # pylint: disable= protected-access
+            if self.policy._state_include_action:
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:245]: i don't see the relationship here -- you test `_state_include_action` and then indiscriminately augment the input with all of the state_info_vars

File: src/garage/tf/algos/npo2.py [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+                k: tf.compat.v1.placeholder(tf.float32,
+                                            shape=[None] * 2 + list(shape),
+                                            name=k)
+                for k, shape in self.policy.state_info_specs
+            }
+            policy_state_info_vars_list = [
+                policy_state_info_vars[k] for k in self.policy.state_info_keys
+            ]
+
+        if self.policy.recurrent:
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:242]: can you remove this from the new policies and just _always_ augment with extra_state_var? if a policy is stateless, it can just have `len(state_info_keys) == 0`

File: src/garage/tf/algos/npo2.py [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+                policy_entropy = pol_dist.entropy()
+
+            # This prevents entropy from becoming negative for small policy std
+            if self._use_softplus_entropy:
+                policy_entropy = tf.nn.softplus(policy_entropy)
+
+            if self._stop_entropy_gradient:
+                policy_entropy = tf.stop_gradient(policy_entropy)
+
+        if not self.policy.recurrent:
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:406]: would this still work if the policy were non-recurrent?

File: src/garage/tf/algos/npo2.py [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+            aug_returns.append(path['returns'])
+        samples_data['rewards'] = np_tensor_utils.pad_tensor_n(
+            aug_rewards, self.max_path_length)
+        samples_data['returns'] = np_tensor_utils.pad_tensor_n(
+            aug_returns, self.max_path_length)
+
+        # Fit baseline
+        logger.log('Fitting baseline...')
+        if hasattr(self.baseline, 'fit_with_samples'):
+            self.baseline.fit_with_samples(paths, samples_data)
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo2.py:455]: we don't have any baselines that implement `fit_with_samples`, so i think you can safely remove this check

File: src/garage/tf/algos/npo2.py (hunk 5 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:86]: ops I copied from the original NPO and wasn't aware. Will fix

File: src/garage/tf/algos/npo2.py (hunk 6 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:126]: Sure

File: src/garage/tf/algos/npo2.py (hunk 8 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:242]: This flag is for applying `boolean_mask` or not. For recurrent, inputs feeding to policies will be [N, T, S^*] while non-recurrent would be (N, S^*)

File: src/garage/tf/algos/npo2.py (hunk 7 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:245]: looks like the comment above is for this block of comment? Yes if we have empty `self.policy.state_info_keys` for non-recurrent policies, we can always augment with `self.policy.state_info_keys`.

File: src/garage/tf/algos/npo2.py (hunk 10 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:455]: Ok

File: src/garage/tf/algos/npo2.py (hunk 9 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:406]: We have tests and benchmarks for NPO with non-recurrent policies.

File: src/garage/tf/algos/__init__.py (hunk 3 above)
Comment (ryanjulian) [reply src/garage/tf/algos/__init__.py:10]: the same goes with all of the primitives -- having DistributionFooPolicy/DistributionFooPolicy2 will just create confusion.

If you don't have the time to run the benchmarks before merging this, then feel free to merge it and send a PR to delete/rename everything later.

File: src/garage/tf/algos/npo2.py (hunk 7 above)
Comment (ryanjulian) [reply src/garage/tf/algos/npo2.py:245]: this is what i'm proposing. i want to delete the property `Policy.recurrent`

File: src/garage/tf/algos/npo2.py (hunk 8 above)
Comment (ryanjulian) [reply src/garage/tf/algos/npo2.py:242]: is there a reason they both can't take the same feed?

File: src/garage/tf/algos/npo2.py (hunk 9 above)
Comment (ryanjulian) [reply src/garage/tf/algos/npo2.py:406]: i'm asking if we can do the same thing for both recurrent and non-recurrent here

**Comment by ryanjulian** [comment]
I am not exactly sure how much work my suggestions about removing `self.policy.recurrent` are -- but if it's a lot of work, you could implement them in #1104 instead.

File: src/garage/tf/algos/npo2.py (hunk 8 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:242]: They both take the same feed -- (N, T, S^), and for non-recurrent policies, since

> Can you repost this as modifying, rather than renaming, NPO/PPO/TRPO? Alternatively, if you're not going to do that in this PR, can you send us a branch which does that?
> 
> This PR makes it difficult to compare your changes between the two versions of the algorithms.

That makes sense. I will repost it.

**Comment by ahtsan** [comment]
> Can you repost this as modifying, rather than renaming, NPO/PPO/TRPO? Alternatively, if you're not going to do that in this PR, can you send us a branch which does that?
> 
> This PR makes it difficult to compare your changes between the two versions of the algorithms.

That makes sense. I will repost it.

File: src/garage/tf/algos/npo2.py (hunk 8 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:242]: Sampler provides trajectories (N, T, S^) to the algorithm, and then passed to the policies. For recurrent policies, the shape matches. For non-recurrent policies, the shape has to be (N, S^) unless we augment it to be (N, 1, S^) to match the shape.

Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:242]: I think it's part of the API in garage that we need to clean up and unify -- Whether we use dense matrix for both recurrent and non-recurrent policies.

Comment (ryanjulian) [reply src/garage/tf/algos/npo2.py:242]: how hard is it to make the non-recurrent policies accept `(N, T, S^)`? -- currently this loss function already computes assuming the inputs are `(N, T, S^)`

Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:242]: Indeed it computes advantage assuming inputs are `(N, T, S^)`, but then filter with valids to make it `(N * [T], S^)`. To make non-recurrent policies accept `(N, T, S^)`, we basically feed a placeholder with such shape, which means whenever the policy accepts data (observation/actions), we need to augment it from `(N, S^)` to `(N, 1, S^)`.

Comment (ryanjulian) [reply src/garage/tf/algos/npo2.py:242]: Okay -- please look at it. If it's too much work, please file an issue about doing this before submitting.

Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:242]: Regarding augmentation with extra_state_var -- What if a recurrent policy with `state_include_action=False`?

Comment (ahtsan) [reply src/garage/tf/algos/npo2.py:242]: I realized in that case `policy_state_info_list` will be empty (since `extra_state_var` is not added if `state_include_action=False`)
